### PR TITLE
execgen: add execgen:instantiate and execgen:let declarations

### PR DESCRIPTION
**Pull request by jordanlewis**
execgen:instantiate provides a way to inform execgen about the possible
inputs to a template function, besides providing a concrete callsite.

This allows us to solve the problem of how to convert a runtime value
into a template value without having to write out a big switch/case
statement by hand.

For example, consider an integer template like this:

    // execgen:template<i>
    func b(i int) {
      switch i {
        case 1:
          fmt.Println("found a 1")
        case 2:
          fmt.Println("found a 2")
        case 3:
          fmt.Println("found a 3")
      }
    }

    func a() {
      b(1)
      b(2)
    }

Previously, this template would be instantiated for all unique callsites
of the template, in this case, 1 and 2.

But what if we had a *runtime* value that we wanted to pass into the `b`
function? How would we do that? Previously, we'd have to have another
trivial case statement, to link up the dynamic value with an
explicit call to the template:

    func a(input int) {
      if input == 1 {
        b(1)
      } else if input == 2 {
        b(2)
      }
    }

Of course this code looks really silly, but without it, we can't
specialize the `b` function properly, since we don't have the
compile-time values 1 and 2 available to do so.

This commit solves this problem by introducing a new directive, called
execgen:instantiate, which provides a list of the possible arguments to
a template function.

With execgen:instantiate, our example can be rewritten like this:

    // execgen:instantiate<1>
    // execgen:instantiate<2>
    // execgen:instantiate<3>
    // execgen:template<i>
    func b(i int) {
      switch i {
        case 1:
          fmt.Println("found a 1")
        case 2:
          fmt.Println("found a 2")
        case 3:
          fmt.Println("found a 3")
      }
    }

    func a(input int) {
        b(input)
    }

Given this input, execgen will generate variants for all of the
instantiate annotations, and create a runtime-to-template switch
function that allows us to simply call b(input), magically turning the
runtime input value into an invocation of a compiled, specialized
function:

    func b_runtime_to_template(i int) {
	    switch i {
	    case 1:
		    b_1()
	    case 2:
		    b_2()
	    case 3:
		    b_3()
	    default:
		    panic(fmt.Sprintf("unknown value %s", i))
	    }
    }

    func a(input int) {
	    b_runtime_to_template(input)
    }

    func b_1() {
	    fmt.Println("found a 1")
    }

    func b_2() {
	    fmt.Println("found a 2")
    }

    func b_3() {
	    fmt.Println("found a 3")
    }

--- 

execgen:let is a new annotation that declares the following slice
literal var as a compile-time argument options list that can be used in
an execgen:instantiate annotation as a shortcut for frequently-used sets
of arguments.

For example, take the following list of execgen:instantiate arguments:

```
// execgen:template<b, i>
// execgen:instantiate<true, 1>
// execgen:instantiate<true, 2>
// execgen:instantiate<true, 3>
// execgen:instantiate<false, 1>
// execgen:instantiate<false, 2>
// execgen:instantiate<false, 3>
func foo (b bool, i int) {
  ...
}
```

As a shortcut to typing out all of these instantiate declarations, we
can do the following with execgen:let:

```
// execgen:let
var bools = []bool{true, false}

// execgen:let
var ints = []int{1, 2, 3}

// execgen:template<b, i>
// execgen:instantiate<bools, ints>
func foo (b bool, i int) {
  ...
}

### CODE REVIEW

**Review by DrewKimball — APPROVED** [review]
Does instantiate work with struct types? If so, it would be nice to have a test for that.

**Comment by jordanlewis** [comment]
If I'm understanding you correctly, instantiate doesn't (or shouldn't) really work with types like structs, since structs can have runtime values inside. The instantiate code really just works with "lexical values" or constants - essentially tokens that are lexicographically comparable. We don't actually check what's inside of the lexical value during execgen compilation, since execgen doesn't actually interpret Go code - it only parses the AST. So if we tried to use anything that wasn't a constant value things would not really make too much sense.

**Comment by DrewKimball** [comment]
> If I'm understanding you correctly, instantiate doesn't (or shouldn't) really work with types like structs, since structs can have runtime values inside. The instantiate code really just works with "lexical values" or constants - essentially tokens that are lexicographically comparable. We don't actually check what's inside of the lexical value during execgen compilation, since execgen doesn't actually interpret Go code - it only parses the AST. So if we tried to use anything that wasn't a constant value things would not really make too much sense.

Ah, I see. I was thinking that struct template variables could be used to handle cases where we don't want to do a cartesian product between the variables with `let` syntax, as in type switching (families vs widths).

**Comment by jordanlewis** [comment]
> Ah, I see. I was thinking that struct template variables could be used to handle cases where we don't want to do a cartesian product between the variables with let syntax, as in type switching (families vs widths).

Yeah, this would be really nice. But unfortunately we're not smart enough in `execgen` (yet?) to handle this - since we don't know how to inspect vars past their lexical values. We could conceivably improve this but it would be significant amounts of work.

TFTR!

bors r+

**Comment by jordanlewis** [comment]
Had to amend the last commit to teach the template outputter to always output in a sorter order, to avoid flakiness due to map iteration order.

File: pkg/sql/colexec/execgen/template.go [hunk 1]
```diff
@@ -401,6 +522,135 @@ func findTemplateFuncs(f *dst.File) map[string]*funcInfo {
...
+	stmtList := make([]dst.Stmt, len(groupedArgs)+1)
+	var i int
+	// Sort firstArgs lexicographically, so we have a consistent output order.
+	firstArgs := make([]string, 0, len(groupedArgs))
+	for firstArg := range groupedArgs {
+		firstArgs = append(firstArgs, firstArg)
+	}
+	sort.Strings(firstArgs)
+
+	for _, firstArg := range firstArgs {
```
Comment (DrewKimball) [inline pkg/sql/colexec/execgen/template.go:635]: [nit] seems like you can just make this `for i, firstArg := range...` and remove the `i` declaration above, right?

**Review by DrewKimball — APPROVED** [review]

File: pkg/sql/colexec/execgen/template.go (hunk 1 above)
Comment (jordanlewis) [reply pkg/sql/colexec/execgen/template.go:635]: Good point, thanks

**Pull request merged**